File: Linguagem GO/Questão7.go
//7. Associação Crie uma classe Escola e uma classe Professor. A associação deve permitir que uma escola tenha vários professores, e um professor possa lecionar em várias escolas.
package main
import "fmt"


type Professor struct {
	nome    string
	escolas []Escola
}

func NewProfessor(nome string) *Professor {
	return &Professor{nome: nome, escolas: []Escola{}}
}

func (p *Professor) addescola(escola *Escola) {
	for _, e := range p.escolas {
		if e.nome == escola.nome {
			return
		}
	}
	p.escolas = append(p.escolas, *escola)
	escola.addprofessor(p)
}

func (p *Professor) ListarEscolas() []string {
	nomes := []string{}
	for _, escola := range p.escolas {
		nomes = append(nomes, escola.nome)
	}
	return nomes
}

type Escola struct {
	nome      string
	professores []Professor
}

func NewEscola(nome string) *Escola {
	return &Escola{nome: nome, professores: []Professor{}}
}

func (e *Escola) addprofessor(professor *Professor) {
	for _, p := range e.professores {
		if p.nome == professor.nome {
			return
		}
	}
	e.professores = append(e.professores, *professor)
	professor.addescola(e)
}

func (e *Escola) ListarProfessores() []string {
	nomes := []string{}
	for _, professor := range e.professores {
		nomes = append(nomes, professor.nome)
	}
	return nomes
}

func main() {
	professor1 := NewProfessor("Dr. Ricardo")
	professor2 := NewProfessor("Prof. Tatyana")

	escola1 := NewEscola("Escola Sesi")
	escola2 := NewEscola("Escola Senhor do Bom Fim")

	escola1.addprofessor(professor1)
	escola1.addprofessor(professor2)
	escola2.addprofessor(professor2)

	fmt.Printf("Professores na %s: %v\n", escola1.nome, escola1.ListarProfessores())
	fmt.Printf("Professores na %s: %v\n", escola2.nome, escola2.ListarProfessores())
	fmt.Printf("Escolas onde %s leciona: %v\n", professor1.nome, professor1.ListarEscolas())
	fmt.Printf("Escolas onde %s leciona: %v\n", professor2.nome, professor2.ListarEscolas())
}

File: Linguagem Python/Questão7.py
#7. Associação Crie uma classe Escola e uma classe Professor. A associação deve permitir que uma escola tenha vários professores, e um professor possa lecionar em várias escolas.

class Professor:
    def __init__(self, nome):
        self.nome = nome
        self.escolas = []

    def addescola(self, escola):
        if escola not in self.escolas:
            self.escolas.append(escola)
            escola.addprofessor(self)

    def listar_escolas(self):
        return [escola.nome for escola in self.escolas]


class Escola:
    def __init__(self, nome):
        self.nome = nome
        self.professores = []

    def addprofessor(self, professor):
        if professor not in self.professores:
            self.professores.append(professor)
            professor.addescola(self)

    def listarprof(self):
        return [professor.nome for professor in self.professores]


professor_1 = Professor("Dr. Ricardo")
professor_2 = Professor("Prof. Tatyana")

escola_1 = Escola("Escola Sesi")
escola_2 = Escola("Escola Senhor do Bom Fim")

escola_1.addprofessor(professor_1)
escola_1.addprofessor(professor_2)
escola_2.addprofessor(professor_2)

print(f"Professores na {escola_1.nome}: {escola_1.listarprof()}")
print(f"Professores na {escola_2.nome}: {escola_2.listarprof()}")

print(f"Escolas onde {professor_1.nome} leciona: {professor_1.listar_escolas()}")
print(f"Escolas onde {professor_2.nome} leciona: {professor_2.listar_escolas()}")
